Build error: undefined reference to `FIPS_mode'

When I build the project using `make`, I encounter the following error.

```
/bin/sh ./libtool --mode=link gcc -o ssh ssh.o readconf.o clientloop.o sshtty.o sshconnect.o sshconnect2.o mux.o ssh-sk-client.o -L. -Lopenbsd-compat/  -Wl,-z,relro -Wl,-z,now -Wl,-z,noexecstack -fstack-protector-strong -pie  -lssh -lopenbsd-compat  /home/user/Applications/gct/gsi/gss_assist/source/libglobus_gss_assist.la /home/user/Applications/gct/gsi/gssapi/source/library/libglobus_gssapi_gsi.la /home/user/Applications/gct/common/source/library/libglobus_common.la  -lcrypto -lmd -lz  -lcrypt -lpthread 
libtool: link: gcc -o .libs/ssh ssh.o readconf.o clientloop.o sshtty.o sshconnect.o sshconnect2.o mux.o ssh-sk-client.o -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-z -Wl,noexecstack -fstack-protector-strong -pie  -L. -Lopenbsd-compat/ -lssh -lopenbsd-compat /home/user/Applications/gct/gsi/gss_assist/source/.libs/libglobus_gss_assist.so /home/user/Applications/gct/callout/source/library/.libs/libglobus_callout.so /home/user/Applications/gct/gsi/gssapi/source/library/.libs/libglobus_gssapi_gsi.so /home/user/Applications/gct/gsi/proxy/proxy_core/source/library/.libs/libglobus_gsi_proxy_core.so /home/user/Applications/gct/gsi/credential/source/library/.libs/libglobus_gsi_credential.so /home/user/Applications/gct/gsi/callback/source/library/.libs/libglobus_gsi_callback.so /home/user/Applications/gct/gsi/callback/source/library/oldgaa/.libs/libglobus_oldgaa.so /home/user/Applications/gct/gsi/cert_utils/source/library/.libs/libglobus_gsi_cert_utils.so /home/user/Applications/gct/gsi/sysconfig/source/library/.libs/libglobus_gsi_sysconfig.so /home/user/Applications/gct/gsi/openssl_module/source/library/.libs/libglobus_openssl.so /home/user/Applications/gct/gsi/proxy/proxy_ssl/source/library/.libs/libglobus_proxy_ssl.so /home/user/Applications/gct/gsi/openssl_error/source/library/.libs/libglobus_openssl_error.so -lssl /home/user/Applications/gct/common/source/library/.libs/libglobus_common.so -lltdl -lcrypto -lmd -lz -lcrypt -lpthread -pthread
/usr/bin/ld: ./libssh.a(kexgexc.o): in function `input_kex_dh_gex_group':
/home/user/Applications/gct/gsi_openssh/source/kexgexc.c:119: undefined reference to `FIPS_mode'
collect2: error: ld returned 1 exit status
```

The issue seems to be that the openssl library on my machine is version 3.0.7. But this merge should have fixed things: https://github.com/gridcf/gct/pull/159 . Is that not so?

If using the latest version of openssl is not possible, I can install version 1.1, but how do I tell `make` or `configure` to use that version?

I think I forgot to include the `fips_mode_replacement.h` header file in https://github.com/gridcf/gct/blob/50fc40ba2eef7197a93807d2391c6dc3bb077a1b/gsi_openssh/source/kexgexc.c (like outlined [here](https://github.com/gridcf/gct/pull/159/files#diff-f4081e66c02bfd7c7de38187912721f6a5e8716daa0e3eec8e08c298dc52bca7)) when updating to GSI-OpenSSH 8.7p1. Which leads to:

```
[...]
/home/user/Applications/gct/gsi_openssh/source/kexgexc.c:119: undefined reference to `FIPS_mode'
[...]
```

Not sure why this doesn't trigger a problem on Rocky Linux 9 or CentOS Stream 9 which both use OpenSSL 3.x already.

Could you manually add the following line after `#include "openbsd-compat/openssl-compat.h"` in the above mentioned `[...]/gsi_openssh/source/kexgexc.c` file:

```
#include "fips_mode_replacement.h"
```

...and try again (but first read the part below)?

> The flags I passed to configure are:
> 
> ```
> ./configure --prefix="/usr/local" --datarootdir="/usr/share" --enable-myproxy --disable-ccommonlibs --disable-gridftp --disable-gram5
> ```

I believe you need the "ccommonlibs" for GSI-OpenSSH to work - at least "libglobus_common" is a dependency for GSI-OpenSSH on CentOS 7. So please clean your build dir, reconfigure without `--disable-ccommonlibs` and try the build again (with the change proposed above).

--- a/gsi_openssh/source/fips_mode_replacement.h
+++ b/gsi_openssh/source/fips_mode_replacement.h
@@ -14,12 +14,18 @@
  * limitations under the License.
  */
 
+#ifndef FIPS_MODE_REPLACEMENT_H
+#define FIPS_MODE_REPLACEMENT_H
+
 #if OPENSSL_VERSION_NUMBER >= 0x30000000L
 /*
- * OpenSSL version 3.0 and up no longer has FIPS_mode().
- * Making a replacement function is not feasible since FIPS would need to be
- * initialized differently in any case.
- * See https://www.openssl.org/docs/manmaster/man7/fips_module.html for details
+ * OpenSSL versions 3.0 and up no longer have FIPS_mode(). To support both
+ * OpenSSL 3.x and older versions for other OSes, we use the replacement
+ * function as shipped by Fedora/RHEL/CentOS in their OpenSSL 3.x packages.
  */
-# define FIPS_mode() 0
-#endif
+# ifndef FIPS_mode
+#  define FIPS_mode() EVP_default_properties_is_fips_enabled(NULL)
+# endif /* FIPS_mode */
+#endif /* openssl */
+
+#endif /* FIPS_MODE_REPLACEMENT_H */

--- a/gsi_openssh/source/kexgex.c
+++ b/gsi_openssh/source/kexgex.c
@@ -34,7 +34,6 @@
 #include <signal.h>
 
 #include "openbsd-compat/openssl-compat.h"
-#include "fips_mode_replacement.h"
 
 #include "sshkey.h"
 #include "cipher.h"

--- a/gsi_openssh/source/kexgexc.c
+++ b/gsi_openssh/source/kexgexc.c
@@ -39,6 +39,7 @@
 #include <signal.h>
 
 #include "openbsd-compat/openssl-compat.h"
+#include "fips_mode_replacement.h"
 
 #include "sshkey.h"
 #include "cipher.h"

